File: LR3/lr3.py
import cv2
import numpy as np

def gauss_func(x, y, a, b, sigma):
    return (1 / (2 * np.pi * sigma ** 2)) * (np.exp(-((x - a) ** 2 + (y - b) ** 2) / (2 * sigma ** 2)))

def build_ker(size, sigma):
    a, b = (size - 1) // 2, (size - 1) // 2
    ker = np.zeros((size, size))
    for i in range(size):
        for j in range(size):
            ker[i, j] = gauss_func(i, j, a, b, sigma)
    ker /= ker.sum()
    return ker

def apply_gaussian_blur(image_path, kernel_size, sigma):
    image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    height = int(image.shape[0] * 0.5)
    width = int(image.shape[1] * 0.5)
    image = cv2.resize(image, (width, height))
    image_copy = image.copy()
    ker = build_ker(kernel_size, sigma)
    center = kernel_size // 2
    for i in range(center, height - center):
        print(i)
        for j in range(center, width - center):
            val = np.sum(image[i - center: i + center + 1, j - center: j + center + 1] * ker)
            image_copy[i, j] = val
    # оригинал
    cv2.imshow('Original Image', image)
    # без cv2
    cv2.imshow('Blurred Image', image_copy)
    # c помощью cv2
    blurred_image = cv2.GaussianBlur(image, (kernel_size, kernel_size), sigma)
    cv2.imshow('Blurred Image with cv2', blurred_image)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

image_path = "D:/704.jpg"
kernel_size = 3
sigma = 2
apply_gaussian_blur(image_path, kernel_size, sigma)

File: LR3/lr3.java
import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.*;
import java.awt.color.ColorSpace;
import java.awt.image.BufferedImage;
import java.awt.image.ColorConvertOp;
import java.io.File;
import java.io.IOException;

public class lr3 {

    public static double gaussFunc(double x, double y, double a, double b, double sigma) {
        double pi = Math.PI;
        return (1 / (2 * pi * Math.pow(sigma, 2)) * (Math.exp(-(Math.pow((x - a), 2) + Math.pow((y - b), 2)) / (2 * Math.pow(sigma, 2)))));
    }

    public static double[][] buildKernel(int size, double sigma) {
        int a = (size - 1) / 2;
        int b = (size - 1) / 2;
        double[][] kernel = new double[size][size];
        double sum = 0;

        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                kernel[i][j] = gaussFunc(i, j, a, b, sigma);
                sum += kernel[i][j];
            }
        }

        // нормализация ядра
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                kernel[i][j] /= sum;
            }
        }
        return kernel;
    }

    public static BufferedImage applyGaussianBlur(BufferedImage image, double[][] kernel) {
        int width = image.getWidth();
        int height = image.getHeight();
        BufferedImage blurredImage = new BufferedImage(width, height, BufferedImage.TYPE_BYTE_GRAY);

        int kernelSize = kernel.length;
        int center = kernelSize / 2;

        for (int i = center; i < height - center; i++) {
            for (int j = center; j < width - center; j++) {
                double red = 0.0;
                double green = 0.0;
                double blue = 0.0;

                for (int m = -center; m <= center; m++) {
                    for (int n = -center; n <= center; n++) {
                        int pixel = image.getRGB(j + n, i + m);
                        double ker = kernel[m + center][n + center];

                        red += ((pixel >> 16) & 0xFF) * ker;
                        green += ((pixel >> 8) & 0xFF) * ker;
                        blue += (pixel & 0xFF) * ker;
                    }
                }
                int blurredColor = ((int) red << 16) | ((int) green << 8) | (int) blue;
                blurredImage.setRGB(j, i, blurredColor);
            }
        }
        return blurredImage;
    }

    public static void main(String[] args) {
        try {
            String imagePath = "D:/704.jpg";
            BufferedImage originalImage = ImageIO.read(new File(imagePath));

            ColorConvertOp op = new ColorConvertOp(ColorSpace.getInstance(ColorSpace.CS_GRAY), null);
            originalImage = op.filter(originalImage, null);

            int kernelSize = 5;
            double sigma = 1.0;

            double[][] kernel = buildKernel(kernelSize, sigma);

            BufferedImage blurredImage = applyGaussianBlur(originalImage, kernel);

            JFrame frame = new JFrame();
            frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

            JPanel panel = new JPanel();
            panel.setLayout(new GridLayout(1, 2));

            panel.add(new JLabel(new ImageIcon(originalImage)));
            panel.add(new JLabel(new ImageIcon(blurredImage)));

            frame.add(panel);
            frame.pack();
            frame.setVisible(true);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}